File: Problems/0100_0199/0173_Binary_Search_Tree_Iterator/Project_Java/BSTIterator.java
import java.util.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class BSTIterator {
    // 16ms
    int index;
    List<Integer> nums;

    public BSTIterator(TreeNode root) {
        nums = new ArrayList<>();
        dfs(root);
        index = 0;
    }
    
    private void dfs(TreeNode node) {
        if (node == null)
            return;
        if (node.left != null)
            dfs(node.left);
        nums.add(node.val);
        if (node.right != null)
            dfs(node.right);
    }

    /** @return the next smallest number */
    public int next() {
        if (index < nums.size())
            return nums.get(index++);
        return -1;
    }
    
    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        if (index < nums.size())
            return true;
        return false;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: Problems/0100_0199/0173_Binary_Search_Tree_Iterator/Project_Go/solution/BSTIterator.go
package solution

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// 40ms

var nums []int

type BSTIterator struct {
	index int
	nums  []int
}

func Constructor(root *TreeNode) BSTIterator {
	nums = make([]int, 0)
	dfs(root)

	BI := new(BSTIterator)
	BI.index = 0
	BI.nums = nums

	return *BI
}

func dfs(node *TreeNode) {
	if node == nil {
		return
	}
	if node.Left != nil {
		dfs(node.Left)
	}
	nums = append(nums, node.Val)
	if node.Right != nil {
		dfs(node.Right)
	}
}

/** @return the next smallest number */
func (this *BSTIterator) Next() int {
	if this.index < len(this.nums) {
		this.index++
		return this.nums[this.index-1]
	}
	return -1
}

/** @return whether we have a next smallest number */
func (this *BSTIterator) HasNext() bool {
	if this.index < len(this.nums) {
		return true
	}
	return false
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Next();
 * param_2 := obj.HasNext();
 */
